### ISSUE TITLE
Build compiles a different number of modules each time

### ISSUE THREAD

**Issue by CarlOlson** [rescript-lang/rewatch#129]
Rerunning build without any code or modification time changes results in compiled modules. Each time it is a different count.

Here are the compiled counts after each run: `1252 -> 600 -> 199 -> 13 -> 30 -> 2 -> 2 -> 2`

<details><summary>Output</summary>

```
[1/7] 📦 Built package tree in 0.07s
[2/7] 🕵️  Found source files in 0.04s
[3/7] 📝 Read compile state 0.11s
[4/7] 🧹 Cleaned 0/2298 0.01s
[5/7] 🧱 Parsed 2 source files in 0.04s
[6/7] ️🌴 Collected deps in 0.14s
[7/7] 🤺 ️Compiled 1252 modules in 15.97s

✨ Finished Compilation in 16.51s

[1/7] 📦 Built package tree in 0.07s
[2/7] 🕵️  Found source files in 0.04s
[3/7] 📝 Read compile state 0.11s
[4/7] 🧹 Cleaned 0/2298 0.01s
[5/7] 🧱 Parsed 2 source files in 0.04s
[6/7] ️🌴 Collected deps in 0.14s
[7/7] 🤺 ️Compiled 600 modules in 11.98s

✨ Finished Compilation in 12.52s

[1/7] 📦 Built package tree in 0.06s
[2/7] 🕵️  Found source files in 0.04s
[3/7] 📝 Read compile state 0.11s
[4/7] 🧹 Cleaned 0/2298 0.01s
[5/7] 🧱 Parsed 1 source files in 0.04s
[6/7] ️🌴 Collected deps in 0.13s
[7/7] 🤺 ️Compiled 199 modules in 4.36s

✨ Finished Compilation in 4.90s

[1/7] 📦 Built package tree in 0.06s
[2/7] 🕵️  Found source files in 0.04s
[3/7] 📝 Read compile state 0.11s
[4/7] 🧹 Cleaned 0/2298 0.01s
[5/7] 🧱 Parsed 1 source files in 0.04s
[6/7] ️🌴 Collected deps in 0.13s
[7/7] 🤺 ️Compiled 13 modules in 0.76s

✨ Finished Compilation in 1.29s

[1/7] 📦 Built package tree in 0.07s
[2/7] 🕵️  Found source files in 0.04s
[3/7] 📝 Read compile state 0.11s
[4/7] 🧹 Cleaned 0/2298 0.01s
[5/7] 🧱 Parsed 1 source files in 0.04s
[6/7] ️🌴 Collected deps in 0.14s
[7/7] 🤺 ️Compiled 30 modules in 1.19s

✨ Finished Compilation in 1.74s

[1/7] 📦 Built package tree in 0.06s
[2/7] 🕵️  Found source files in 0.04s
[3/7] 📝 Read compile state 0.11s
[4/7] 🧹 Cleaned 0/2298 0.01s
[5/7] 🧱 Parsed 1 source files in 0.04s
[6/7] ️🌴 Collected deps in 0.14s
[7/7] 🤺 ️Compiled 2 modules in 0.08s

✨ Finished Compilation in 0.61s

[1/7] 📦 Built package tree in 0.07s
[2/7] 🕵️  Found source files in 0.04s
[3/7] 📝 Read compile state 0.12s
[4/7] 🧹 Cleaned 0/2298 0.01s
[5/7] 🧱 Parsed 1 source files in 0.04s
[6/7] ️🌴 Collected deps in 0.14s
[7/7] 🤺 ️Compiled 2 modules in 0.12s

✨ Finished Compilation in 0.68s

[1/7] 📦 Built package tree in 0.07s
[2/7] 🕵️  Found source files in 0.04s
[3/7] 📝 Read compile state 0.12s
[4/7] 🧹 Cleaned 0/2298 0.01s
[5/7] 🧱 Parsed 1 source files in 0.04s
[6/7] ️🌴 Collected deps in 0.14s
[7/7] 🤺 ️Compiled 2 modules in 0.09s

✨ Finished Compilation in 0.64s
```

</details>

**Comment by rolandpeelen** [comment]
@jfrolich -- I think I ran into this at some point but haven't been able to recreate. @CarlOlson -- Which version were you running at the time? Has it been fixed since by any chance?

**Comment by CarlOlson** [comment]
> @jfrolich -- I think I ran into this at some point but haven't been able to recreate. @CarlOlson -- Which version were you running at the time? Has it been fixed since by any chance?

Sorry, I finally got around to trying this again. The issue is gone in version `1.0.9`. I do get a similar issue when some warnings trigger, but I believe it would be fixed by #120.

**Comment by CarlOlson** [comment]
Actually I can recreate the issue:

1. Run `rescript build` in workspace of a mono repo that has a `type: dev` folder
2. Run `rewatch build` in the project root

So I think this is more of an issue with unexpected data in the `lib` folder. This should get fixed by #124.

For now, running `clean` before switching between `rescript build` and `rewatch build` is enough.

**Comment by rolandpeelen** [comment]
Ok - check. I'm commenting in the other issue - I did do some work on the `type: dev` stuff, and I now have some more context here.
